### PR TITLE
wdio-lambdatest-service: Added new service to update job status

### PR DESCRIPTION
**Pull request by kanhaiya15**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

- Using lambdatest service, for updating job status (passed/failed) 
- It help to run tunnel for local testing if capability has tunnel property to true.  

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-cli/src/constants.js [hunk 1]
```diff
@@ -100,6 +101,7 @@ export const QUESTIONNAIRE = [{
     choices: [
         'On my local machine',
         'In the cloud using Sauce Labs',
+        'In the cloud using LambdaTest',
```
Comment (christian-bromann) [inline packages/wdio-cli/src/constants.js]: Given that the only question specific to LambdaTest is username and key, we can move this to the section:
```
In the cloud using Browserstack, LambdaTest or Testingbot or a different service
```

File: examples/cloudservices/lambdatest.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+(async () => {
+    const browser = await remote({
+        capabilities: {
+            platform : 'Windows 10',
+            browserName : 'Chrome',
+            version : '79.0',
+            resolution : '1024x768'
+        },
+        hostname: 'hub.lambdatest.com',
+        port: 80,
```
Comment (christian-bromann) [inline examples/cloudservices/lambdatest.js]: Can we prefer SSL usage here?

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            }),
+            new Promise((resolve, reject) => {
+                timer = setTimeout(function() {
+                    reject(
+                        'LambdaTest Tunnel failed to start within 60 seconds!'
+                    )
+                }, 60000)
+            })
+        ]).then(
+            function(result) {
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: I would prefer arrow functions: `then((result) => {`

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                clearTimeout(timer)
+                performance.mark('ltTunnelEnd')
+                performance.measure('bootTime', 'ltTunnelStart', 'ltTunnelEnd')
+                return Promise.resolve(result)
+            },
+            function(err) {
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: same

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            typeof this.lambdatestTunnelProcess.isRunning !== 'function' ||
+            !this.lambdatestTunnelProcess.isRunning()
+        ) {
+            return
+        }
+
+        let timer
+        return Promise.race([
+            new Promise((resolve, reject) => {
+                this.lambdatestTunnelProcess.stop(err => {
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: Can we use `promisify` here?

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            }),
+            new Promise((resolve, reject) => {
+                timer = setTimeout(
+                    () =>
+                        reject(
+                            new Error(
+                                'LambdaTest Tunnel failed to stop within 60 seconds!'
+                            )
+                        ),
+                    60000
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: this should be defined in a constant file to stay consistent with the codebase

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            new Promise((resolve, reject) => {
+                this.lambdatestTunnelProcess.stop(err => {
+                    if (err) {
+                        return reject(err)
+                    }
+                    resolve()
+                })
+            }),
+            new Promise((resolve, reject) => {
+                timer = setTimeout(
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: how about just having
```suggestion
                timer = setTimeout(() => reject(new Error(...)), TUNNEL_STOP_TIMEOUT)
```

File: packages/wdio-lambdatest-service/src/launcher.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                    () =>
+                        reject(
+                            new Error(
+                                'LambdaTest Tunnel failed to stop within 60 seconds!'
+                            )
+                        ),
+                    60000
+                )
+            })
+        ]).then(
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/launcher.js]: how about:
```js
]).then(() {
    clearTimeout(timer)
    return Promise.resolve()
}, (err) {
    clearTimeout(timer)
    return Promise.reject(err)
})
```

File: packages/wdio-lambdatest-service/src/service.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    // gather information about runner
+    beforeSession(config, capabilities) {
+        this.config = config
+        this.capabilities = capabilities
+        const lambdaCredentials = {
+            username: this.config.user,
+            accessKey: this.config.key
+        }
+        this.isServiceEnabled =
+            lambdaCredentials.username && lambdaCredentials.accessKey
```
Comment (christian-bromann) [inline packages/wdio-lambdatest-service/src/service.js]: no new line needed

**Comment by kanhaiya15** [comment]
Hi @christian-bromann ,
I understand your concern about build slowness but we get this feature request from lots of users including other providers, they moved to our platform and they don't want to change their test scripts so that we need this feature.I am working on your suggestions , will push this changes ASAP.

File: packages/wdio-cli/src/constants.js (hunk 1 above)
Comment (kanhaiya15) [reply packages/wdio-cli/src/constants.js]: Done

File: examples/cloudservices/lambdatest.js (hunk 2 above)
Comment (kanhaiya15) [reply examples/cloudservices/lambdatest.js]: Done

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 3 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Done

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 4 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Done

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 5 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Due to single parameter in callback of stop method, getting error by using promisify

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 6 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Done

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 7 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Done

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 8 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/launcher.js]: Done

File: packages/wdio-lambdatest-service/src/service.js (hunk 9 above)
Comment (kanhaiya15) [reply packages/wdio-lambdatest-service/src/service.js]: Done

**Comment by kanhaiya15** [comment]
Hi @christian-bromann ,
Pushed your suggestions changes. Please let me know if missed anything and I really appreciate your suggestion, I have added this to my checklist.
Thanks

File: packages/wdio-lambdatest-service/src/launcher.js (hunk 5 above)
Comment (christian-bromann) [reply packages/wdio-lambdatest-service/src/launcher.js]: ah yeah, promisify doesn't work here

**Comment by christian-bromann** [comment]
@kanhaiya15 any chance you were able to add more unit tests?

**Comment by kanhaiya15** [comment]
Hi @christian-bromann ,
Sorry for delayed response, actually I was on leave.For this service we are going to maintain in our own organisation.
Thnaks for your help & support
Regards
Kanhaiya

**Comment by christian-bromann** [comment]
@kanhaiya15 alright! Feel free to add the service to our [service list](https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-cli/src/constants.js#L67-L88) so it is still picked up by the config wizard and documented on webdriver.io!

**Pull request closed without merging**